src/Common/ColumnsHashing.h

```diff
@@ -383,37 +384,75 @@ struct HashMethodSerialized
             for (auto row_size : row_sizes)
                 total_size += row_size;
 
-            serialized_buffer.resize(total_size);
-
-            const size_t rows = row_sizes.size();
-            char * memory = serialized_buffer.data();
-            std::vector<char *> memories(rows);
-            serialized_keys.resize(rows);
-            for (size_t i = 0; i < row_sizes.size(); ++i)
+            use_batch_serialize = shouldUseBatchSerialize();
+            if (use_batch_serialize)
             {
-                memories[i] = memory;
-                serialized_keys[i].data = memory;
-                serialized_keys[i].size = row_sizes[i];
+                serialized_buffer.resize(total_size);
 
-                memory += row_sizes[i];
-            }
+                const size_t rows = row_sizes.size();
+                char * memory = serialized_buffer.data();
+                std::vector<char *> memories(rows);
+                serialized_keys.resize(rows);
+                for (size_t i = 0; i < row_sizes.size(); ++i)
+                {
+                    memories[i] = memory;
+                    serialized_keys[i].data = memory;
+                    serialized_keys[i].size = row_sizes[i];
 
-            for (size_t i = 0; i < keys_size; ++i)
-            {
-                if constexpr (nullable)
-                    key_columns[i]->batchSerializeValueIntoMemoryWithNull(memories, null_maps[i]);
-                else
-                    key_columns[i]->batchSerializeValueIntoMemory(memories);
+                    memory += row_sizes[i];
+                }
+
+                for (size_t i = 0; i < keys_size; ++i)
+                {
+                    if constexpr (nullable)
+                        key_columns[i]->batchSerializeValueIntoMemoryWithNull(memories, null_maps[i]);
+                    else
+                        key_columns[i]->batchSerializeValueIntoMemory(memories);
+                }
             }
         }
     }
 
+    bool shouldUseBatchSerialize() const
+    {
+#if defined(__aarch64__)
+        // On ARM64 architectures, always use batch serialization for performance.
```

Comment:
But why?

Reply:
It is from my observation from perf tests on x86 and arm.